engine/src/main/java/org/hibernate/validator/internal/engine/groups/ExpandedGroup.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.engine.groups;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ * Represents a validation group and all the groups it extends ("group inheritance").
+ *
+ * @author Gunnar Morling
+ */
+public class ExpandedGroup implements Iterable<Group> {
```

Comment:
I tried that first actually, but it didn't work out eventually. In the validator algorithm we still need individual groups to check constraint presence; I still think it'd work in some way, but it'd need more investigation and changes which I tried to minimize as far as reasonably possible for now as the fix needs backporting to 4.3. That said, I am all for revisting the entire implementation at some point. I got some ideas, but it'll be a larger piece of work I am afraid.